packages/headless/src/api/generated-answer/generated-answer-client.ts

```diff
@@ -0,0 +1,166 @@
+import {fetchEventSource} from '@microsoft/fetch-event-source';
+import {Logger} from 'pino';
+import {SearchAppState} from '../..';
+import {AsyncThunkOptions} from '../../app/async-thunk-options';
+import {ClientThunkExtraArguments} from '../../app/thunk-extra-arguments';
+import {GeneratedAnswerErrorPayload} from '../../features/generated-answer/generated-answer-actions';
+import {URLPath} from '../../utils/url-utils';
+import {resetTimeout} from '../../utils/utils';
+import {SearchAPIClient} from '../search/search-api-client';
+import {
+  GeneratedAnswerCitationsPayload,
+  GeneratedAnswerMessagePayload,
+  GeneratedAnswerPayloadType,
+  GeneratedAnswerStreamEventData,
+} from './generated-answer-event-payload';
+import {GeneratedAnswerStreamRequest} from './generated-answer-request';
+
+export interface GeneratedAnswerAPIClientOptions {
+  logger: Logger;
+}
+
+export interface AsyncThunkGeneratedAnswerOptions<
+  T extends Partial<SearchAppState>
+> extends AsyncThunkOptions<
+    T,
+    ClientThunkExtraArguments<SearchAPIClient, GeneratedAnswerAPIClient>
+  > {}
+
+const buildStreamingUrl = (url: string, orgId: string, streamId: string) =>
+  new URLPath(
+    `${url}/rest/organizations/${orgId}/machinelearning/streaming/${streamId}`
+  );
+
+const MAX_RETRIES = 3;
+const MAX_TIMEOUT = 5000;
+const EVENT_STREAM_CONTENT_TYPE = 'text/event-stream';
+export const RETRYABLE_STREAM_ERROR_CODE = 1;
+
+class RetryableError extends Error {}
+class FatalError extends Error {}
+
+interface DispatchActions {
+  updateMessage: (payload: GeneratedAnswerMessagePayload) => void;
+  updateCitations: (payload: GeneratedAnswerCitationsPayload) => void;
+  updateError: (payload: GeneratedAnswerErrorPayload) => void;
+  setIsLoading: (isLoading: boolean) => void;
+  resetAnswer: () => void;
+}
```

Comment:
Instead of taking 5 actions, I think we should:
1. Just like the search API client, not use redux terminology (e.g., actions).
2. Take similar callbacks to a [`WriteableStream`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStream#examples) (as if `streamGeneratedAnswer` was the equivalent of calling `someStream.pipeTo`).
  * write(chunk)
  * close()
  * abort()

Ideally, I would've preferred to use the `ReadableStream` and `WriteableStream` directly, but they aren't supported by Node 12, which our docs say we support.

Reply:
I can change the terminology for sure.
I'm not sure I understand how you envision the client working like a writeable stream. `fetchEventSource` already works in a way that resembles it. but there's a lot of stuff happening around it that IMO doesn't make sense to handle in the redux action over right here in the client.

Reply:
Maybe i'm just not understanding your suggestion.

Reply:
I agree it's odd to pass in essentially redux actions but it's hard to find a context that has both access to redux and the mechanics of the stream. I could move most of the logic out of here and into the action but then the client doesn't really have much reason for existing as I think it'll basically be empty. Here it at least encapsulated the streaming complexity away from redux definitions.